eng/common/scripts/artifact-metadata-parsing.ps1

```diff
@@ -0,0 +1,469 @@
+    $attempts += 1
+  }
+}
+
+# Parse out package publishing information given a maven POM file
+function ParseMavenPackage($pkg, $workingDirectory) {
+  [xml]$contentXML = Get-Content $pkg
+
+  $pkgId = $contentXML.project.artifactId
+  $pkgVersion = $contentXML.project.version
+  $groupId = if ($contentXML.project.groupId -eq $null) { $contentXML.project.parent.groupId } else { $contentXML.project.groupId }
+
+  # if it's a snapshot. return $null (as we don't want to create tags for this, but we also don't want to fail)
+  if ($pkgVersion.Contains("SNAPSHOT")) {
+    return $null
+  }
+
+  $releaseNotes = &"${PSScriptRoot}/../Extract-ReleaseNotes.ps1" -ChangeLogLocation @(Get-ChildItem -Path $pkg.DirectoryName -Recurse -Include "$($pkg.Basename)-changelog.md")[0]
+
+  $readmeContentLoc = @(Get-ChildItem -Path $pkg.DirectoryName -Recurse -Include "$($pkg.Basename)-readme.md")[0]
+  if (Test-Path -Path $readmeContentLoc) {
+    $readmeContent = Get-Content -Raw $readmeContentLoc
+  }
+
+  return New-Object PSObject -Property @{
+    PackageId      = $pkgId
+    PackageVersion = $pkgVersion
+    Deployable     = $forceCreate -or !(IsMavenPackageVersionPublished -pkgId $pkgId -pkgVersion $pkgVersion -groupId $groupId.Replace(".", "/"))
+    ReleaseNotes   = $releaseNotes
+    ReadmeContent  = $readmeContent
+  }
+}
+
+# Returns the maven (really sonatype) publish status of a package id and version.
+function IsMavenPackageVersionPublished($pkgId, $pkgVersion, $groupId) {
+  try {
+
+    $uri = "https://oss.sonatype.org/content/repositories/releases/$groupId/$pkgId/$pkgVersion/$pkgId-$pkgVersion.pom"
+    $pomContent = Invoke-RestMethod -Method "GET" -Uri $uri
+
+    if ($pomContent -ne $null -or $pomContent.Length -eq 0) {
+      return $true
+    }
+    else {
+      return $false
+    }
+  }
+  catch {
+    $statusCode = $_.Exception.Response.StatusCode.value__
+    $statusDescription = $_.Exception.Response.StatusDescription
+
+    # if this is 404ing, then this pkg has never been published before
+    if ($statusCode -eq 404) {
+      return $false
+    }
+
+    Write-Host "VersionCheck to maven for packageId $pkgId failed with statuscode $statusCode"
+    Write-Host $statusDescription
+    exit(1)
+  }
+}
+
+# make certain to always take the package json closest to the top
+function ResolvePkgJson($workFolder) {
+  $pathsWithComplexity = @()
+  foreach ($file in (Get-ChildItem -Path $workFolder -Recurse -Include "package.json")) {
+    $complexity = ($file.FullName -Split { $_ -eq "/" -or $_ -eq "\" }).Length
+    $pathsWithComplexity += New-Object PSObject -Property @{
+      Path       = $file
+      Complexity = $complexity
+    }
+  }
+
+  return ($pathsWithComplexity | Sort-Object -Property Complexity)[0].Path
+}
+
+# Parse out package publishing information given a .tgz npm artifact
+function ParseNPMPackage($pkg, $workingDirectory) {
+  $workFolder = "$workingDirectory$($pkg.Basename)"
+  $origFolder = Get-Location
+  mkdir $workFolder
```

Comment:
As we start to try and reuse these scripts we should try and switch to use the cmdlets directly because alias' like these often times do different things on different OS's. Some OS's have a built in command others use the cmdlet and they generally have different logic and options, so we should try to always use the cmdlet if possible.

Reply:
I don't understand what you mean by this comment. `Get-Location` IS the cmdlet versus `pwd` which would be the local alias right?

Reply:
For this particular comment I was looking at `mkdir` which on windows is a alias to the cmdlet but on linux it uses the shell native linux command and they have differences. Namely one of them will create subdirectories and the other will not, which has bitten me in the past (see https://github.com/Azure/azure-sdk-tools/pull/357 for some examples).  To be clear this isn't something you need to change with this PR but as we refactor theses scripts an expect them to run on multiple OS's we will start to hit these issues.

Reply:
Ah. Ok you meant `mkdir`. Yeah you're absolutely right. Gotta fix that.

Reply:
@weshaggard this is fixed. now. All `mkdir` have been converted over to `New-Item` calls with `-Force` attached to ensure that missing parents are created.